ext/opentelemetry-instrumentation-fastapi/src/opentelemetry/instrumentation/fastapi/__init__.py

```diff
@@ -0,0 +1,82 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Optional
+
+import fastapi
+from starlette.routing import Match
+
+from opentelemetry.ext.asgi import OpenTelemetryMiddleware
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.instrumentation.fastapi.version import __version__  # noqa
+
+
+class FastAPIInstrumentor(BaseInstrumentor):
+    """An instrumentor for FastAPI
+
+    See `BaseInstrumentor`
+    """
+
+    _original_fastapi = None
+
+    @staticmethod
+    def instrument_app(app: fastapi.FastAPI):
+        """Instrument a previously instrumented FastAPI application.
+        """
+        if not getattr(app, "is_instrumented_by_opentelemetry", False):
+            app.add_middleware(
+                OpenTelemetryMiddleware,
+                span_details_callback=_get_route_details,
+            )
+            app.is_instrumented_by_opentelemetry = True
+
+    def _instrument(self, **kwargs):
+        self._original_fastapi = fastapi.FastAPI
+        fastapi.FastAPI = _InstrumentedFastAPI
+
+    def _uninstrument(self, **kwargs):
+        fastapi.FastAPI = self._original_fastapi
+
+
+class _InstrumentedFastAPI(fastapi.FastAPI):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.add_middleware(
+            OpenTelemetryMiddleware, span_details_callback=_get_route_details
+        )
+
+
+def _get_route_details(scope):
+    """Callback to retrieve the starlette route being served.
+
+    TODO: there is currently no way to retrieve http.route from
+    a starlette application from scope.
+
+    See: https://github.com/encode/starlette/pull/804
+    """
+    app = scope["app"]
+    route = None
+    for starlette_route in app.routes:
+        match, _ = starlette_route.matches(scope)
+        if match == Match.FULL:
```

Comment:
Can you briefly explain what this "hack" does to get the http.route? What is in app.routes?

Reply:
app.routes is the list of routes added to starlette: https://www.starlette.io/routing/

This mimics logic in the starlette app router itself to match to one of the routes in the list, and find the appropriate one.

In the future the code in the linked pr in the docstring will add the appropriate objects in the scope, at which point we will be able to extract the 'http.route' object by querying the first object in the "routes" value added to the scope: https://github.com/encode/starlette/pull/804.

By doing this, we can avoid two iterations of all match objects, which will lead to performance hits depending on the complexity and number of routes added.

Reply:
Right that makes sense. What I'm confused about is, if `starlette_route` is one of the list of routes added, why match it with scope? Does scope contain the actual route that is being used, but we don't know the key? Also, is `matches` a method of the `Route` object in starlette? I can't seem to find documentation on it.

Reply:
yeah, the documentation around starlette is a bit rough. I found it via looking through the code (particularly the router part. the starlette PR above is still a good reference).

a starlette_route is just an object, and you iterate through all routes mounted on the app, regardless of whether it's appropriate for the http route or not. the scope contains the url and path, which is what the route is matching against (as well as other things like HTTP method).